src/andromede/simulation/optimization.py

```diff
@@ -408,25 +273,128 @@ def register_connection_fields_expressions(
+                self, component_id: str, name: str
+            ) -> IndexingStructure:
+                return (
+                    network.get_component(component_id).model.parameters[name].structure
+                )
+
+            def get_parameter_structure(self, name: str) -> IndexingStructure:
+                raise RuntimeError("Component context should have been initialized.")
+
+            def get_variable_structure(self, name: str) -> IndexingStructure:
+                raise RuntimeError("Component context should have been initialized.")
+
+        return Impl()
+
+    def expand_operators(self, expression: ExpressionNode) -> ExpressionNode:
+        dimensions = ProblemDimensions(self.block_length(), self.scenarios)
+        time_bound_evaluator = self.evaluate_time_bound
+        return expand_operators(
+            expression,
+            dimensions,
+            time_bound_evaluator,
+            self._indexing_structure_provider,
+        )
+
+    def _make_parameter_getter(self) -> ParameterGetter:
+        ctxt = self
+
+        class Impl(ParameterGetter):
+            def get_parameter_value(
+                self,
+                component_id: str,
+                parameter_name: str,
+                timestep: Optional[int],
+                scenario: Optional[int],
+            ) -> float:
+                return _get_parameter_value(
+                    ctxt,
+                    timestep,
+                    scenario,
+                    component_id,
+                    parameter_name,
+                )
+
+        return Impl()
+
+    def linearize_expression(
+        self,
+        expanded: ExpressionNode,
+        timestep: Optional[int] = None,
+        scenario: Optional[int] = None,
+    ) -> LinearExpression:
+        return linearize_expression(
+            expanded, timestep, scenario, self._parameter_getter
+        )
 
+    def compute_indexing(self, expression: ExpressionNode) -> IndexingStructure:
+        return compute_indexation(expression, self._indexing_structure_provider)
 
-def _compute_indexing_structure(
-    context: ComponentContext, constraint: Constraint
+    def _make_constant_value_provider(self) -> ValueProvider:
+        """
+        Value provider which only provides values for constant parameters
+        """
+        context = self
+        network = self.network
+
+        class Impl(ValueProvider):
+            def get_component_variable_value(
+                self, component_id: str, name: str
+            ) -> float:
+                raise NotImplementedError(
+                    "Cannot provide variable value at problem build time."
+                )
+
+            def get_component_parameter_value(
+                self, component_id: str, name: str
+            ) -> float:
+                model = network.get_component(component_id).model
+                structure = model.parameters[name].structure
+                if structure.time or structure.scenario:
```

Comment:
Actually with the current implementation, we perform the "operators" expansion only once for all scenarios and timesteps. So we really need to have only one value of bounds for all scenarios and timesteps.

But I agree it's probably a constraint  which is too strong. To allow different values for different scenarios, we will need to perform operators expansion once for each scenario. This may hurt a little performances for cases where we build a problem for multiple scenarios, but it's not an issue right now. Also, at this point, we may decided to analyse the actual time/scenario dependencies of bounds to decide if we need to do operators expansion only once for all, or once for each timestep/scenario.

For now I propose to leave it this way, and rework that particular point when needed.